src/Control/Monad/Freer/Internal.hs

```diff
@@ -194,162 +184,27 @@ sendM = send
 --   'Data.Function.&' runEff2 eff2Arg1 eff2Arg2
 --   'Data.Function.&' 'run'
 -- @
-run :: Eff '[] a -> a
-run (Val x) = x
-run _       = error "Internal:run - This (E) should never happen"
+run :: Eff '[Identity] a -> a
+run = runIdentity . runM
+{-# INLINE run #-}
 
 -- | Like 'run', 'runM' runs an 'Eff' computation and extracts the result.
 -- /Unlike/ 'run', 'runM' allows a single effect to remain within the type-level
 -- list, which must be a monad. The value returned is a computation in that
 -- monad, which is useful in conjunction with 'sendM' or 'liftBase' for plugging
 -- in traditional transformer stacks.
 runM :: Monad m => Eff '[m] a -> m a
-runM (Val x) = return x
-runM (E u q) = case extract u of
-  mb -> mb >>= runM . qApp q
-  -- The other case is unreachable since Union [] a cannot be constructed.
-  -- Therefore, run is a total function if its argument terminates.
-
--- | Like 'replaceRelay', but with support for an explicit state to help
--- implement the interpreter.
-replaceRelayS
-  :: s
-  -> (s -> a -> Eff (v ': effs) w)
-  -> (forall x. s -> t x -> (s -> Arr (v ': effs) x w) -> Eff (v ': effs) w)
-  -> Eff (t ': effs) a
-  -> Eff (v ': effs) w
-replaceRelayS s' pure' bind = loop s'
-  where
-    loop s (Val x)  = pure' s x
-    loop s (E u' q) = case decomp u' of
-        Right x -> bind s x k
-        Left  u -> E (weaken u) (tsingleton (k s))
-      where
-        k s'' x = loop s'' $ qApp q x
-{-# INLINE replaceRelayS #-}
-
--- | Interpret an effect by transforming it into another effect on top of the
--- stack. The primary use case of this function is allow interpreters to be
--- defined in terms of other ones without leaking intermediary implementation
--- details through the type signature.
-replaceRelay
-  :: (a -> Eff (v ': effs) w)
-  -> (forall x. t x -> Arr (v ': effs) x w -> Eff (v ': effs) w)
-  -> Eff (t ': effs) a
-  -> Eff (v ': effs) w
-replaceRelay pure' bind = loop
-  where
-    loop (Val x)  = pure' x
-    loop (E u' q) = case decomp u' of
-        Right x -> bind x k
-        Left  u -> E (weaken u) (tsingleton k)
-      where
-        k = qComp q loop
-{-# INLINE replaceRelay #-}
-
-replaceRelayN
```

Comment:
I haven't reimplemented this in `too-fast-too-free` but it seems niche enough that we probably don't need to provide it?

Reply:
Nevermind. This is pretty useful! I'll reimplement it before merging.